### ISSUE TITLE
OutOfMemoryError when mixing usage of 0.13.0 and 0.13.1-RC2

### ISSUE THREAD

**Issue by paulp** [sbt/sbt#979]
As described here: https://groups.google.com/forum/#!topic/sbt-dev/NPb0kZ2rpsY

Ideally you can reproduce it with
`git clone https://github.com/paulp/sbtbug && cd sbtbug && ./runner.sh`
However there is something contributing non-determinism, and sometimes it works fine.

Okay, I found the likely source of non-determinism. "sbt run" is behaving very differently in 0.13.1-RC2. It doesn't wait for main to finish. If I use this main body:

``` scala
  def main(args: Array[String]): Unit = {
    println("a")
    Thread.sleep(10000);
    println("b")
  }
```

Under 0.13.0 it prints a, sleeps 10 seconds, prints b. Under 0.13.1-RC2 it prints a and then exits.

0.13.0:

```
[info] Running bippy.Main
a
b
[success] Total time: 12 s, completed Nov 12, 2013 7:29:42 PM
```

0.13.1-RC2:

```
[info] Running bippy.Main
a
[success] Total time: 1 s, completed Nov 12, 2013 7:29:58 PM
```

**Comment by harrah** [comment]
I appreciate you tracking down that non-determinism, since that's a regression that will also need to get fixed.

**Comment by paulp** [comment]
It becomes slipperier when I try to isolate that. But I do have an attachment which should make it easy to at least see the memory exhaustion.

**Comment by paulp** [comment]
Or maybe not, it seems I can't attach a tgz. But I'll keep it in case reproducing it is hard.

**Comment by harrah** [comment]
Ok, thanks.  I have a guess that I'll try tomorrow.

**Comment by harrah** [comment]
I can't reproduce either issue.  I ran that main several times using 0.13.1-RC2 and it always waited for `run` to finish.  I don't know if it matters, but I use a plain startup path and not sbt-extras.

**Comment by harrah** [comment]
Ah, I can reproduce the awt and the not waiting for `run` issues on a mac, though.

**Comment by SethTisue** [comment]
I just recently started getting mysterious OutOfMemoryErrors when starting sbt, and I also recently started testing 0.13.1-RC2, so perhaps I've been experiencing this too. if there's anything I should do next time it happens, let me know. (On the group @harrah mentioned something about a cache? What cache do I try clearing to see if the error goes away?)

**Comment by harrah** [comment]
The cache is at `target/streams/compile/incCompileSetup/`.  It will be versioned for RC3, so I don't think you need to try to reproduce it.

**Comment by paulp** [comment]
On Thu, Nov 14, 2013 at 5:54 AM, Mark Harrah notifications@github.comwrote:

> The cache is at target/streams/compile/incCompileSetup/. It will be
> versioned for RC3, so I don't think you need to try to reproduce it..

Since it doesn't take too much trying on my part, I tried. Happily, removing

```
project/target/streams/compile/incCompileSetup
```

... does indeed makes the problem go away. I assume this is the path you
meant - the path you gave also exists, but does not affect the issue.

**Comment by harrah** [comment]
In that case, it is the cache for compiling the build definition.  If sbt can't get past the build definition, it won't ever read the cache for the project itself.  (Sorry, I didn't actually notice it was failing on startup and not `compile`.)

**Comment by harrah** [comment]
The main issue was fixed by #988.  I'll open new issues to track the other problems.
